### PR TITLE
add scaling factors to stableswap pool creation

### PR DESCRIPTION
**Pull request by catShaark**
Closes: #2166

## What is the purpose of the change

add scaling factors to stableswap pool creation

## Brief Changelog

- add scaling factors to `MsgCreateStableswapPool`
- add scaling factors to `NewStableswapPool`
- add scaling factors checking in `MsgCreateStableswapPool` `ValidateBasic`
- make changes to existing tests accordingly
- add new error type : `ErrInvalidScalingFactors` 

## Testing and Verifying

This change is already covered by existing tests

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes 
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? yes
  - How is the feature or change documented? no

### CODE REVIEW

File: x/gamm/pool-models/stableswap/msgs.go [hunk 1]
```diff
@@ -51,6 +53,11 @@ func (msg MsgCreateStableswapPool) ValidateBasic() error {
 	} else if len(msg.InitialPoolLiquidity) > 2 {
 		return types.ErrTooManyPoolAssets
 	}
+	if len(msg.ScalingFactors) != 2 {
+		return types.ErrInvalidScalingFactors
+	} else if msg.ScalingFactors[0] == 0 || msg.ScalingFactors[1] == 0 {
+		return types.ErrInvalidScalingFactors
+	}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/msgs.go:60]:
```suggestion
	// valid scaling factor lengths are 0, or one factor for each asset
	if len(msg.ScalingFactors) != 0 && len(msg.ScalingFactors) != len(msg.InitialPoolLiquidity) {
		return types.ErrInvalidScalingFactors
	}
```

**Review by ValarDragon — APPROVED** [review]

File: x/gamm/pool-models/stableswap/msgs.go (hunk 1 above)
Comment (catShaark) [reply x/gamm/pool-models/stableswap/msgs.go:60]: Shoudn't we also check for zero scaling factor

Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/msgs.go:60]: 0 scaling factor should get inferred to just be all 1's

**Comment by catShaark** [comment]
@ValarDragon, I've commited your suggestion, is this ready to merge ?

File: x/gamm/pool-models/stableswap/pool.go [hunk 2]
```diff
@@ -20,13 +20,20 @@ var _ types.PoolI = &Pool{}
 // * len(initialLiquidity) = 2
 // * FutureGovernor is valid
 // * poolID doesn't already exist
-func NewStableswapPool(poolId uint64, stableswapPoolParams PoolParams, initialLiquidity sdk.Coins, futureGovernor string) (Pool, error) {
+func NewStableswapPool(poolId uint64, stableswapPoolParams PoolParams, initialLiquidity sdk.Coins, scalingFactors []uint64, futureGovernor string) (Pool, error) {
+	if len(scalingFactors) == 0 {
+		scalingFactors = []uint64{1, 1}
+	} else if scalingFactors[0] == 0 || scalingFactors[1] == 0 {
+		scalingFactors = []uint64{1, 1}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:27]:
```suggestion
		return Pooll{}, errors.New("one of the scaling factors provided was 0")
```

Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/pool.go:27]: (reverts back to your prior commit, I think that was right)

Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/pool.go:27]: If this was on the osmosis main repo, I'd just commit this suggestion and then hit merge. After this is updated, I think its good to merge.

File: x/gamm/pool-models/stableswap/pool.go [hunk 3]
```diff
@@ -20,13 +20,20 @@ var _ types.PoolI = &Pool{}
 // * len(initialLiquidity) = 2
 // * FutureGovernor is valid
 // * poolID doesn't already exist
-func NewStableswapPool(poolId uint64, stableswapPoolParams PoolParams, initialLiquidity sdk.Coins, futureGovernor string) (Pool, error) {
+func NewStableswapPool(poolId uint64, stableswapPoolParams PoolParams, initialLiquidity sdk.Coins, scalingFactors []uint64, futureGovernor string) (Pool, error) {
+	if len(scalingFactors) == 0 {
+		scalingFactors = []uint64{1, 1}
+	} else if scalingFactors[0] == 0 || scalingFactors[1] == 0 {
+		return Pooll{}, types.ErrInvalidStableswapScalingFactors
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/pool.go:27]:
```suggestion
		return Pool{}, types.ErrInvalidStableswapScalingFactors
```

**Comment by ValarDragon** [comment]
Yeah I think so!

When these branches get onto the main osmosis repo, will be so much faster to get merged :)

**Comment by ValarDragon** [comment]
We can merge once the test passes, sorry for this crazy number of minor back and forths :sweat:

**Comment by catShaark** [comment]
> We can merge once the test passes, sorry for this crazy number of minor back and forths :sweat: 

Sorry I forgot about the tests, I got them all passed now

**Pull request merged**